crates/service/src/middleware/inject_allocation.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use alloy::primitives::Address;
+use axum::{
+    extract::{Request, State},
+    middleware::Next,
+    response::Response,
+};
+use tap_core::receipt::SignedReceipt;
+use thegraph_core::DeploymentId;
+use tokio::sync::watch;
+
+/// The current query Allocation Id address
+#[derive(Clone)]
+pub struct Allocation(pub Address);
+
+impl From<Allocation> for String {
+    fn from(value: Allocation) -> Self {
+        value.0.to_string()
+    }
+}
+
+/// State to be used by allocation middleware
+#[derive(Clone)]
+pub struct AllocationState {
+    /// watcher that maps deployment ids to allocation ids
+    pub deployment_to_allocation: watch::Receiver<HashMap<DeploymentId, Address>>,
+}
+
+/// Injects allocation id in extensions
+/// - check if allocation id already exists
+/// - else, try to fetch allocation id from deployment_id to allocations map
+///
+/// Requires signed receipt Extension to be added OR deployment id
+pub async fn allocation_middleware(
+    State(my_state): State<AllocationState>,
+    mut request: Request,
+    next: Next,
+) -> Response {
+    if let Some(receipt) = request.extensions().get::<SignedReceipt>() {
+        let allocation = receipt.message.allocation_id;
+        request.extensions_mut().insert(Allocation(allocation));
+    } else if let Some(deployment_id) = request.extensions().get::<DeploymentId>() {
+        if let Some(allocation) = my_state
```

Comment:
Shoudn`t it log when allocation is not found, or either the same for `SignedReceipt` or `DeploymentId`?, for example
```
// adding this to the end of the ifs
         else {
            tracing::warn!("No allocation found for deployment: {:?}", deployment_id);
        }
    } else {
        tracing::warn!("Neither SignedReceipt nor DeploymentId was found in request extensions.");
    }
```